docs/Customization/Keys-Traits-Configs.rst

```diff
@@ -0,0 +1,95 @@
+
+Top-level traits specify that the Top has been parameterized to read some custom Key and optionally instantiate and connect a widget defined by that Key. Traits **should not** mandate the instantiation of custom logic. In other words, traits should be written with ``CanHave`` semantics, where the default behavior when the Key is unset is a no-op.
+
+Top-level traits should be defined and documented in subprojects, alongside their corresponding Keys. The traits should then be added to the Top being used by Chipyard.
+
+Below we see the traits for the GCD example. The Lazy trait connects the GCD module to the Diplomacy graph, while the Implementation trait causes the Top to instantiate an additional port and concretely connect it to the GCD module.
+
+.. literalinclude:: ../../generators/example/src/main/scala/GCD.scala
+    :language: scala
+    :start-after: DOC include start: GCD lazy trait
+    :end-before: DOC include end: GCD imp trait
+
+These traits are added to the default ``Top`` in Chipyard.
+
+.. literalinclude:: ../../generators/example/src/main/scala/Top.scala
+    :language: scala
+    :start-after: DOC include start: Top
+    :end-before: DOC include end: Top
+
+Mixins
+------
+
+Mixins set the keys to a non-default value. Together, the collection of Mixins which define a configuration generate the values for all the keys used by the generator.
+
+For example, the ``WithGCDMixin`` is parameterized by the type of GCD widget you want to instantiate. When this mixin is added to a config, the ``GCDKey`` is set to a instance of ``GCDParams``, informing the previously mentioned traits to instantiate and connect the GCD widget appropriately.
+
+.. literalinclude:: ../../generators/example/src/main/scala/ConfigMixins.scala
+    :language: scala
+    :start-after: DOC include start: GCD mixin
+    :end-before: DOC include end: GCD mixin
+
+We can use this mixin when composing our configs.
+
+.. literalinclude:: ../../generators/example/src/main/scala/RocketConfigs.scala
+    :language: scala
+    :start-after: DOC include start: GCDTLRocketConfig
+    :end-before: DOC include end: GCDTLRocketConfig
+
+
+BuildTop
+--------
+
+The ``BuildTop`` key is special, because sometimes, we need to instantiate TestHarness modules to interface with a custom widget. The ``BuildTop`` key provides a function which can call some method of the Top to instantiate these TestHarness modules. Since the ``BuildTop`` key is called from the TestHarness, these modules will appear in the TestHarness. The config system also lets the ``BuildTop`` key look recursively into previous definitions of itself. This enables composability of the Top configurations.
```

Comment:
As you can see from my previous suggestions, I think `Top` should be in "code" format. I also think that `TestHarness` should be the same. Basically, I say that anything that is present in the code should be "code format".

Reply:
I would double check that this is done everywhere.